controllers/userController.js

```diff
@@ -0,0 +1,43 @@
+const keywordModel = require("../models/keywordModel");
+const userModel = require("../models/userModel");
+const { isValidString } = require("../utils/validation");
+
+const create = async (req, res) => {
+  if (!isValidString(req.body.uid)) {
+    return res.status(400).send({ message: "[InvalidUid] Error occured" });
+  }
+  if (!isValidString(req.body.email)) {
+    return res.status(400).send({ message: "[InvalidEmail] Error occured" });
+  }
+  if (!isValidString(req.body.displayName)) {
+    return res.status(400).send({ message: "[InvalidDisplayName] Error occured" });
+  }
+  if (!isValidString(req.body.photoURL)) {
+    return res.status(400).send({ message: "[InvalidPhotoURL] Error occured" });
+  }
+
+  const { uid, email, displayName, photoURL } = req.body;
+  const isNewUser = (await userModel.find({ uid })).length === 0;
+
+  try {
+    if (isNewUser) {
+      const userCreated = await userModel.create({ uid, email, displayName, photoURL });
+
+      return res.status(201).json({ userCreated });
+    } else {
+      const userUpdated = await userModel.findOneAndUpdate(
+        { uid },
+        { $set: { lastSignInAt: new Date() } },
+        { new: true }
+      );
+
+      return res.status(201).json({ userUpdated });
+    }
```

Comment:
```suggestion
    const userInfo = await userModel.findOneAndUpdate(
      { uid },
      { 
        email,
        displayName,
        photoURL
      },
      { upsert: true }
    );

      return res.status(201).json({ userInfo });
    }
```
`findOneAndUpdate`의 `upsert` 옵션을 사용하면, 코드 길이를 더 줄일 수 있다고 생각합니다.
그리고 `lastSignInAt`는 `updatedAt`의 이름을 변경한 것으로 동작은 그래도 적용되어, update 쿼리 적용시에 자동으로 시간도 변경됩니다! 
- [mongoose의 findOneAndUpdate 쿼리](https://mongoosejs.com/docs/tutorials/findoneandupdate.html)
- [mongoose의 timestamp](https://mongoosejs.com/docs/timestamps.html)

Reply:
```suggestion
    const userUpserted = await userModel.findOneAndUpdate(
      { uid },
      {
        $setOnInsert: {
          email,
          displayName,
          photoURL,
        },
      },
      { upsert: true, new: true }
    );
    return res.status(201).json({ userInfo });
```

의견 감사해요 한솔님. 추가로 조사해보니, upsert일 때만 업데이트를 도와주는 `$setOnInsert` 연산자가 있었어요. 요걸로 한솔 의견대로 코드 가독성 높이고, 제가 우려했던 점도 보완할 수 있었어요. 😃

앞서 제안 주신 `upsert`를 사용하면 `lastSignInAt` 외 다른 데이터의 업데이트도 불가피했어요. `email`, `displayName` 등 불변성을 지켜야할 데이터를 변경하는 것이라 분기 처리로 구현을 결정했었습니다.

이때 `$setOnInsert`가 유용했는데요. 이 메소드는 MongoDB 연산자이고, 필드값을 업데이트하는 `$set`과 비슷합니다. 다만, upsert의 경우에만 작동하게 해요. 즉 (유저 정보가 이미 있어서) upsert가 일어나지 않으면, 아무 동작도 하지 않아서 업데이트가 발생하지 않습니다. 저희 의도 대로 동작하면서도, `lastSignInAt(updatedAt)`도 업데이트 할 수 있었어요.

참고로, MongoDB 공식 문서를 보면 `$set`과도 함께 사용되고 있습니다. 앞으로 연산자를 활용할 일이 많은데, 자주 사용할 문법들은 함께 스터디 해요!⚡

```js
db.products.updateOne(
  { _id: 1 },
  {
     $set: { item: "apple" },
     $setOnInsert: { defaultQty: 100 }
  },
  { upsert: true }
)
```


- 출처: [MongoDB - $setOnInsert](https://www.mongodb.com/ko-kr/docs/manual/reference/operator/update/setOnInsert/)

Reply:
다만 추가로 궁금한 건, 데이터가 실제 업데이트 되지는 않은 것인데 `lastSignInAt(updatedAt)`이 어떻게 업데이트 된걸까요? 가입이 아닌 로그인 케이스에서도 (실제 업데이트는 없었지만) `findOneAndUpdate`를 호출했기 때문일까요? 아래 설명을 참고했어요.

> Mongoose will then set createdAt when the document is first inserted, and update updatedAt whenever you update the document using save(), updateOne(), updateMany(), findOneAndUpdate(), update(), replaceOne(), or bulkWrite().

- 출처 [mongoose - timestamps](https://mongoosejs.com/docs/timestamps.html)

Reply:
`$setOnInsert` 앞으로 다른 기능 개발할 때도 유용하게 많이 사용될 것 같네요!
제 의견 반영을 위해서 더 리서치해주셔서 감사해요! 🙇‍♀️

하단의 진태님 의문에 대해서 덕분에 저도 메타인지 되었네요!
해당 내용에 대해서 리서치 해봤는데, 진태님이 올려주신 공식문서 설명과 함께 하단의 "Disabling Timestamps" 파트에서 

> save(), updateOne(), updateMany(), findOneAndUpdate(), update(), replaceOne(), and bulkWrite() all support a timestamps option. Set timestamps: false to skip setting timestamps for that particular operation.

라고 되어 있는걸 보면, `findOneAndUpdate()` 쿼리를 사용하면 무조건 `updateAt`은 업데이트되는 것으로 저는 이해했습니다. 
그래서 `$setOnInsert`를 사용해서 업데이트된 데이터가 없지만, `findOneAndUpdate()` 쿼리를 사용했기 때문에 `updatedAt`의 시간이 업데이트되는 것 같습니다.

Reply:
@jin-ttao 덕분에 `findOneAndUpdate` 와 upsert 옵션을 함께 쓸 때, `insert`와 `update`의 상황 구분을 어떻게 할지 방법을 알게 되었네요. 자세한 설명 감사합니다. 

결국, `insert`시에는 `$setOnInsert`에 적힌 대로 update가 일어나고, `filter`에 있는 것이 발견되어 `update` 할 시에는 lastSignInAt만 업데이트 되는 군요. 재밌네요 :)

Reply:
@SsongQ-92 @only-pine 
- 결정사항: `create`, `update` 등 DB 변경의 결과는 `___Result` 혹은`___ListResult` 변수명으로 한다.

감사합니다! 구두로 논의한대로 `findOneAndUpdate`의 반환값의 변수명만 아래 처럼 추가로 수정했습니다.
```js
const userResult = await userModel.findOneAndUpdate(
// ...
);
```